Update survival tp53/telomerase analysis (#1276)
AlexsLemonade/OpenPBTA-analysis: analyses/survival-analysis/survival-analysis_tp53_telomerase.Rmd

  # convert to years
  mutate(OS_years = OS_days / 365.25)

# Show the data we will use for analysis
meta_indep
```
 
#### Create subsets of data for different models

We create two new tibbles from `meta_indep`:

+ `meta_indep_hgg` which contains a new column `hgg_group` that contains levels (in order) `non-HGAT` and `HGAT` indicating whether it's an HGG sample
+ `meta_indep_cancers` which is subsetted to only certain cancer groups of interest that have suitable sample sizes which we find here.

```{r hgg_split}
# Define HGG groups
hgg_cancer_groups <- c("High-grade glioma astrocytoma", "Diffuse midline glioma", "Diffuse intrinsic pontine glioma") 
hgg_levels <- c("non-HGAT", "HGAT")

# separate into HGG or non-HGG and make the new column a factor
meta_indep_hgg <- meta_indep %>% 
   mutate(hgg_group = ifelse(cancer_group %in% hgg_cancer_groups,
                             hgg_levels[2], # second is HGAT 
                             hgg_levels[1]), # first is non-HGAT
          hgg_group = factor(hgg_group, levels = hgg_levels) # ensures non-HGAT is the reference level
   )
```

```{r find_cancer_groups_to_use}
# Define cancer groups of interest with higher Ns

> # We want to only include groups that have at least THREE in EITHER category. 

Author: ```suggestion
# We want to only include groups that have at least THREE in the DECEASED category. 
```

Reviewer: good catch!

# Find them here:
cancer_groups_of_interest_df <- meta_indep %>%
  drop_na() %>%
  count(cancer_group,OS_status) %>%
  spread(OS_status, n,

>          # any missing values (indicating 0 deceased or 0 living) should be 0, not NA

Reviewer 2: Is this comment correct? I am confused.

Reviewer: The comment is explaining line 147 - `fill = 0`

Reviewer: This part of the code is still counting the number of events, and `NA`s will occur when spreading if there is no recorded event. `indicating 0 deceased or 0 living` is referring to number of participants, not the recoding.

I'll make sure this is doing what I think it is and rephrase a little.

Reviewer 2: My issue is with the phrasing. Using "no events" seems less confusing to me.

         fill = 0) %>%
  filter(DECEASED >= 3) 
# show result: there are 7 groups to keep
cancer_groups_of_interest_df

# Subset to only the cancer groups with potentially suitable sample size
meta_indep_cancer_groups <- meta_indep %>%
  filter(cancer_group %in% cancer_groups_of_interest_df$cancer_group) %>%
  # set reference level to Medulloblastoma
  mutate(cancer_group = fct_relevel(cancer_group, "Medulloblastoma"))

# How many patients got removed?
nrow(meta_indep) - nrow(meta_indep_cancer_groups)
```

## Univariate Models

 
Here, we run a survival analyses for several _independent_ predictors, one at a time. We use cox regression for all analyses (including categorical), as demonstrated in `survival-analysis_template.Rmd`.

+ `tp53_score` (continuous)
+ `telomerase_score` (continuous)
+ `hgg_group` (categorical)

> + `cancer_group` (categorical)

Author: One thing we also might want to do just as a gut check is run a univariate analysis on molecular subtype for `hgg_group` and `cancer_group` here which have subtypes. This can be in this notebook or can be a part of the immune deconv updates..

Reviewer: I think since the immune deconv is already looking at molecular signatures an extent, we can do it there.

Author: sounds good!

### TP53 

```{r cox_tp53}
fit_save_model(meta_indep, 
              "tp53_score", 
             file.path(results_dir, "cox_terms_tp53.RDS"))

> 

Author: all of these RDS files appear NULL when reading them in...

Reviewer: Oh no, I will have a look..

Reviewer: This was a silly typo! Fix incoming!

               file.path(results_dir, "cox_terms_cancer_group.RDS"))
```

### Cancer group (log rank)

Here, we perform a log-rank analysis specifically for the `cancer_group` predictor, using data that only considers the 7 cancer groups determined to have suitable sample sizes (`DECEASED >= 3`). The output below shows:

+ Table of model coefficients
+ Matrix of `Pairwise comparisons using Log-Rank test`, which shows which cancer groups are significantly different from one another

Note that confidence intervals are so large that survival plots are not clearly legible, so we don't make that visualization here.

```{r cancer_group_log_rank}

# Print model coefficients
fit_save_model(meta_indep_cancer_groups,
               "cancer_group",
               file.path(results_dir,
                         "log_rank_survival_per_cancer_group.RDS"),
               test = "log.rank")

# Compare pairwise with Bonferroni correction for multiple testing
survminer::pairwise_survdiff(survival::Surv(time = OS_years, event = OS_status) ~ cancer_group,
                             meta_indep_cancer_groups %>% mutate(OS_status = ifelse(OS_status == "LIVING", 0, 1)), 

>                              p.adjust.method = "bonferroni")

Reviewer: The `survival_analysis()` function was originally created to do the recoding within the function and expects "LIVING" and "DECEASED" - 
https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/51e4fca33eaa306ec7e63291217dc03551bb1e13/analyses/survival-analysis/util/survival_models.R#L97 

I suppose the code would still work if this were already recoded, but I thought it would be preferable to follow with the precedent that was set in the [template Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/51e4fca33eaa306ec7e63291217dc03551bb1e13/analyses/survival-analysis/survival-analysis_template.Rmd) not to recode.

>
>_Hazard ratios._ The exponentiated coefficients (exp(coef) = exp(-0.53) = 0.59), also known as hazard ratios, give the effect size of covariates. For example, being female (sex=2) reduces the hazard by a factor of 0.59, or 41%. Being female is associated with good prognostic.
>
>_Confidence intervals of the hazard ratios._ The summary output also gives upper and lower 95% confidence intervals for the hazard ratio (exp(coef)), lower 95% bound = 0.4237, upper 95% bound = 0.816.
>
>_Global statistical significance of the model._ Finally, the output gives p-values for three alternative tests for overall significance of the model: The likelihood-ratio test, Wald test, and score logrank statistics. These three methods are asymptotically equivalent. For large enough N, they will give similar results. For small N, they may differ somewhat. The Likelihood ratio test has better behavior for small sample sizes, so it is generally preferred.
>
> Below, HR = exp(coeff) 

### Interaction model: `tp53_score*telomerase_score*hgg_group` 

> Interaction terms are generally not signifcant, except `tp53_score:telomerase_score` which is only marginally significant at `P=0.0358`. This result suggests an additive model is more appropriate.

Reviewer 2: If it's easy enough to extract just the p-value from what `fit_save_model()`, you could put this after the code chunk and use  p = `r <variable with the p-value in it>`. Just a bit of future-proofing for thought.

Reviewer: Done.

```{r multi_interaction_3terms}
fit_save_model(meta_indep_hgg,    
             "tp53_score*telomerase_score*hgg_group",
             file.path(results_dir, "cox_interaction_terms_tp53_telomerase_hgg.RDS"),
             "multivariate"
            )

  
```

### Additive model: `tp53_score+telomerase_score+hgg_group` 

```{r multi_additive_3terms}
fit_save_model(meta_indep_hgg,    
             "tp53_score+telomerase_score+hgg_group",
             file.path(results_dir, "cox_additive_terms_tp53_telomerase_hgg.RDS"),
             "multivariate")
```

> ### Additive models: ``tp53_score+telomerase_score` _for each cancer group_

Reviewer 2: ```suggestion
### Additive models: `tp53_score+telomerase_score` _for each cancer group_
```

> Note the model for DIPG does not converge!

> > ONLY the EPN model shows significant effects for both `tp53_score` and `telomerase_score`, although CIs are fairly wide. The DPG model is marginally significant for `tp53_score` only.

Author: ```suggestion
> ONLY the EPN model shows significant effects for both `tp53_score` and `telomerase_score`, although CIs are fairly wide. The DMG model is marginally significant for `tp53_score` only.
```

Author: This is good (re: DMG), and expected!

#### TP53 score * Telomerase score * HGG group

```{r multivar-fit-3terms}
fit_plot_save_cox_multivariate(meta_indep_hgg, 
                          "tp53_score*telomerase_score*hgg_group", 
                          results_dir)
```

#### TP53 score * Telomerase score, for each cancer group.

```{r multivar-fit-2terms}
for (cg in cancer_groups_of_interest_df$cancer_group) {
  cat("==============", cg, "==================")
  # filter to cancer group of interest
  fit_data <- cancer_groups_of_interest_df %>% 
    filter(cancer_group == cg)
  
  # Define filename where model will be saved
  filename <- paste0("cox_reg_results_per_tp53_score-by-telomerase_score_",
                     stringr::str_replace_all(cg, " ","-"),
                     ".tsv")
  
  # Fit model

>   fit_plot_save_cox_multivariate(meta_indep, 

Reviewer 2: Pretty sure this is the source of the bug @jharenza mentioned

```suggestion
  fit_plot_save_cox_multivariate(fit_data, 
```

Author: updated this, but also was a bug above, too
